### DISCUSSION TITLE
Memory is not being freed?

### DISCUSSION THREAD

**Discussion by Muzosh** [lima-vm/lima#2720] [Q&A]
I ran some memory intensive task today and it maxed out at 8GB (which was defined in my config). I was this in MacOS Activity Monitor next to a "Virtual Machine Service for limactl" process.

However, once those memory intensive tasks finished, the process stayed at 8GB indefinitely (until colima restart), slowing down my host Mac. 

Is this intended behavior? I would expect the process to jump down to some idle value (around ~800-1000MB after fresh start).

**Comment by jandubois** [comment]
Is this using QEMU or VZ?

I know qemu has a "ballooning" device that can be configured to release memory back to the OS, but I don't know if that is supported on macOS. Either way, Lima doesn't use it.

So the best way to return memory to the OS is to stop your VMs when you don't need/use them.

**Comment by Muzosh** [comment]
Yea it’s VZ. Thanks for the info. On 10. 10. 2024, at 8:40 PM, Jan Dubois ***@***.***> wrote:﻿
Is this using QEMU or VZ?
I know qemu has a "ballooning" device that can be configured to release memory back to the OS, but I don't know if that is supported on macOS. Either way, Lima doesn't use it.
So the best way to return memory to the OS is to stop your VMs when you don't need/use them.

**Reply by jandubois** [reply]
It looks like for VZ we are using a ballooning memory device, so I think it should have worked. Maybe file a GitHub issue about it?

**Reply by Muzosh** [reply]
Let me see if I can reproduce it and collect all details.

**Comment by okch-codes** [comment]
Same problem here. I use VZ for the Rosetta emulation.

**Reply by lmachineone** [reply]
Was this ever solved?

**Reply by jandubois** [reply]
There is no change. I did not follow this topic closely, but I was under the impression that even if you use the ballooning device to reclaim memory, it will only be available to other VM instances and not be returned to macOS for general usage. This may have changed in macOS 26; I don't know.

There is an open PR #4828, which claims to somehow work around this issue, but later comments indicate that the macOS issue is still unresolved. And even then there is the problem of reviewing a single PR with over 4000 added lines...

**Reply by afbjorklund** [reply]
* https://github.com/lima-vm/lima/issues/4220

**Comment by lmachineone** [comment]
Same here, I use VZ without rosetta, macbook pro m4

It accumulates memory, and never cleans it, I need to run colima restart or colima stop --force to fix it.
Like, right now my containers consume around 1.5gb, plus base ubuntu image, should stop at 2.5gb max. But it's consuming 7.21 GB of Memory, 8.61 GB of VM Compressed, and 2.31GB of Real Memory

Like, see images below:

-----

# before running colima restart (clean without --force)

<img width="1322" height="750" alt="image" src="https://github.com/user-attachments/assets/fbf3a122-f56d-45a7-8777-dcc3d3fe1329" />

-----

# And after colima restart:

<img width="1313" height="735" alt="image" src="https://github.com/user-attachments/assets/af6aa30d-caf6-4d07-8b08-82f923ed90c3" />

------

Compressed memory wen't down 1.65 GB,
4.05 GB -> 2.40 GB

------

Swap wen't down even more, 1.92 GB
4.71 FB -> 2.79 GB

------

Would this be a colima's configuration / code issue.
Or lima?

Because I use colima with lima.

**Reply by afbjorklund** [reply]
As far as I know, it is an OS issue. The memory can been ballooned (hidden) in the guest, but it is not released to the host

**Reply by okch-codes** [reply]
[This is](https://orbstack.dev/blog/dynamic-memory) how OrbStack resolved it, just in case.

**Reply by msimkunas** [reply]
> [This is](https://orbstack.dev/blog/dynamic-memory) how OrbStack resolved it, just in case.

This blog post doesn't really go into much detail though. The implementation details are not publicly shared for obvious reasons.

**Reply by afbjorklund** [reply]
We have only tested with QEMU/hvf and Vz, there _might_ be more options available to [libkrun](https://github.com/containers/libkrun) (eventually) and the VMM

* https://github.com/lima-vm/lima/pull/4226
